### ISSUE TITLE
Could not compile examples on Mac OS X

### ISSUE THREAD

**Issue by zonyitoo** [hkuplg/fcore#58]
I just clone the latest version on Github, and I still could not compile some examples, such as `CallByName.sf`.

Here is the output

``` bash
$ f2j -rv CallByName.sf
CallByName using [Naive]
Compiling to Java source code ( ./CallByName.java )
f2j: sorry! (unimplemented feature or known bug)
    BaseTransCFJava.trans.JNew: no idea how to do
```

**Comment by bixuanzju** [comment]
@zhiyuanshi George, I looked at the dumped code

```
let something = \ (_ : Unit).
                  {java.lang.System.out.println("something called")
                  ;1}
in
let callByName = \ (x : Unit -> Int).
                   {java.lang.System.out.println(x)
                   ;java.lang.System.out.println(x)
                   ;java.lang.System.out.println(x)
                   ;"Leaving callByName"}
in
{java.lang.System.out.println("* callByName *")
;callByName
   (\ (_ : Unit). something ())}
```

The argument `x` of `callByName` is a function, that's why java method call failed. Is there something wrong in the front-end?

**Comment by zonyitoo** [comment]
Not only `CallByName.sf`, but also the following:
- `ChainingMethods.sf`

```
$ f2j -r ChainingMethods.sf
ChainingMethods using [Naive]
Compiling to Java source code ( ./ChainingMethods.java )
error: equals is not a member of the type char
```
- `Id.sf`

```
$ f2j -r Id.sf
Id using [Naive]
Compiling to Java source code ( ./Id.java )
f2j: Parse error before tokens:
    [Tforall,Tupperid "A",Tdot,Tupperid "A",Tarrow,Tupperid "A",Tcparen,Tlowerid "id",Tjavaclass "java.lang.Integer",Tint 10]
```
- `Issue_185.sf`

```
$ f2j -r Issue_185.sf                                                                                                                                    ⏎
Issue_185 using [Naive]
Compiling to Java source code ( ./Issue_185.java )
[1]    27675 segmentation fault  f2j -r Issue_185.sf
```
- `LazyAdd.sf`

```
$ f2j -r LazyAdd.sf
LazyAdd using [Naive]
Compiling to Java source code ( ./LazyAdd.java )
./LazyAdd.java:17: 错误: 二元运算符 '+' 的操作数类型错误
            final java.lang.Integer x7 = x3 + x6;
                                            ^
  第一个类型:  Closure
  第二个类型: Closure
1 个错误
错误: 找不到或无法加载主类 LazyAdd
rm: *.class: No such file or directory
```

My OS's locale is set to Simplify Chinese. The message means that there is a type error on binary operator '+', the first operand is a Closure and the second one is also a Closure. I think the problem may be located in the runtime library.
- `ObjectAlgebar.sf`

```
$ f2j -r ObjectAlgebra.sf
ObjectAlgebra using [Naive]
Compiling to Java source code ( ./ObjectAlgebra.java )
f2j: Parse error before tokens:
    [Tlowerid "f",Tcparen,Toparen,Tlowerid "f",Tdot,Tlowerid "lit",Tint 2,Tcparen,Tsemi,Tlet,Tlowerid "merge",Tupperid "A",Tupperid "B",Toparen,Tlowerid "alg1",Tcolon,Tupperid "ExpAlg",Tupperid "A",Tcparen,Toparen,Tlowerid "alg2",Tcolon,Tupperid "ExpAlg",Tupperid "B",Tcparen,Teq,Tocurly,Tlowerid "lit",Teq,Tlam,Toparen,Tlowerid "x",Tcolon,Tjavaclass "java.lang.Integer",Tcparen,Tdot,Tlowerid "alg1",Tdot,Tlowerid "lit",Tlowerid "x",Tmerge,Tlowerid "alg2",Tdot,Tlowerid "lit",Tlowerid "x",Tcomma,Tlowerid "add",Teq,Tlam,Toparen,Tlowerid "x",Tcolon,Tupperid "A",Tandtype,Tupperid "B",Tcparen,Tdot,Tlam,Toparen,Tlowerid "y",Tcolon,Tupperid "A",Tandtype,Tupperid "B",Tcparen,Tdot,Tlowerid "alg1",Tdot,Tlowerid "add",Tlowerid "x",Tlowerid "y",Tmerge,Tlowerid "alg2",Tdot,Tlowerid "add",Tlowerid "x",Tlowerid "y",Tccurly,Tin,Tlet,Tlowerid "newAlg",Teq,Tlowerid "merge",Tjavaclass "java.lang.Integer",Tjavaclass "java.lang.String",Tlowerid "subEvalAlg",Tlowerid "printAlg",Tin,Toparen,Tlam,Toparen,Tlowerid "x",Tcolon,Tjavaclass "java.lang.String",Tcparen,Tdot,Tlowerid "x",Tcparen,Toparen,Tlowerid "exp1",Toparen,Tjavaclass "java.lang.Integer",Tandtype,Tjavaclass "java.lang.String",Tcparen,Tlowerid "newAlg",Tcparen]
```

and the same problem in compiling `ObjectAlgebra*.sf`, `StateMonadList.sf`, `StringLength.sf`, `Visitor.sf`, `datatype.sf`, `test.sf` and `test2.sf`.

I haven't test those examples in the secondly directories.

Am I really using the latest compiler? Which branch should I use?

**Comment by zhiyuanshi** [comment]
Remember to work on the `develop` branch. The `master` branch is quite old. (We should fix this to make it something like a stable release though.)

The call by name feature is still under development. So if you see the source file refers to something like `lazy`, `byName`, there's a high chance that it does not work.

`examples/Issue_185.sf` produces a segfault, which is a known issue.

For the rest, you see a parser error because there is a recent change to the syntax of the source language. Namely, type arguments in expressions should be enclosed in square brackets. If you've added the square brackets and the code still won't compile, do report again here.

**Comment by bixuanzju** [comment]
Also related to #59

**Comment by zhiyuanshi** [comment]
Fix most programs by commit 15e7192 and open new issues (#60, #61, #62, #63) for the rest which are assignable to others.
